proposals/0243-character-literals.md

```diff
@@ -0,0 +1,107 @@
+
+ - syntax for [multi-line string literals](https://github.com/apple/swift-evolution/blob/master/proposals/0168-multi-line-string-literals.md) uses triple quotes (`"""`)
+
+ - string interpolation syntax uses standard double quote syntax. 
+
+ - raw-mode string literals settled into the `#""#` syntax. 
+
+ - In current discussions around [regex literals](https://forums.swift.org/t/string-update/7398/6), most people seem to prefer slashes (`/`) or `#//#` syntax.
+
+ Given that, and the desire for lightweight syntax for single character syntax, and the precedent in other languages for characters, it is natural to use single quotes for this purpose.
+
+## Detailed design
+
+ This is a change that is internal to the Swift compiler and does not affect how these literal values are represented at runtime and hence does not affect the ABI. Single quoted literals are largely identical to double quoted `String` literals, supporting the same existing escape syntax, and they reuse the same code in the lexer which happened to already support parsing single quoted syntax. However, the compiler would in addition perform a best-effort attempt at validating that they contain a single extended grapheme cluster at compile time, as it currently does when an `as Character` type coercion annotation is present. Validation behaviour for `Unicode.Scalar` literals will be unaffected. 
+ 
+```Swift
+// Modified String literal protocol hierarchy:
+ExpressibleByStringLiteral
+  ↳ ExpressibleByExtendedGraphemeClusterLiteral
+      ↳ ExpressibleByUnicodeScalarLiteral
+          ↳ @_marker ExpressibleBySingleQuotedLiteral
+              ↳ @_marker ExpressibleByASCIILiteral
+```
+This is realised by introducing two new `ExpressibleBy` marker protocols: `ExpressibleBySingleQuotedLiteral` and `ExpressibleByASCIILiteral` which are inserted above the existing `ExpressibleByUnicodeScalarLiteral` in the double quoted literal protocols. As they are prefixed with `@_marker` it is assumed this will not affect the ABI of the existing protocol's witness table layouts. The `ExpressibleBySingleQuotedLiteral` is used only to change the default type of single quoted literals in an expression without type context and the `ExpressibleByASCIILiteral` used to gate the ASCII to integer value conversions. While the inheritance relationship of the ABI-locked `ExpressibleBy` protocols technically entails that `Character` and `Unicode.Scalar` literals can be implicitly promoted to `String` literals, it would be possible in future for the compiler to statically reject such cases at the type checking stage, without affecting ABI, in the interest of untangling the various textual literal forms. As literal delimiters are a purely compile-time construct, and all double-quoted literals currently default to `String`, this will have zero impact on all existing Swift code.
```

Comment:
let’s get a compiler dev to review this so we don’t have to say “it is assumed this will not affect the ABI”

also: “technically entails” is grammatically weird